pkg/proxy/proxy.go

```diff
@@ -339,14 +337,26 @@ func uploadToBlobStorage(ctx context.Context, cacheID int) (*DockerGHAUploadCach
 			return nil, fmt.Errorf("no presigned URL found")
 		}
 
-		client, err := azblob.NewClientWithNoCredential(cacheEntry.BackendReserveResponse.AzureBlob.PreSignedURL, nil)
-		if err != nil {
-			return nil, fmt.Errorf("failed to create Azure Blob client: %w", err)
-		}
+		for attempt := 0; attempt < maxRetries; attempt++ {
+			contentReader := bytes.NewReader(finalBuffer.Bytes())
+			req, err := http.NewRequestWithContext(ctx, http.MethodPut, cacheEntry.BackendReserveResponse.AzureBlob.PreSignedURL, contentReader)
```

Comment:
`contentReader` is not needed. You can directly pass `*bytes.Buffer` as a body.

```suggestion
			req, err := http.NewRequestWithContext(ctx, http.MethodPut, cacheEntry.BackendReserveResponse.AzureBlob.PreSignedURL, &finalBuffer)
```

Reply:
This needs testing in S3 and Azure both. Will pick it up later